prov/shm/src/smr_ep.c

```diff
@@ -0,0 +1,747 @@
+	.listen = fi_no_listen,
+	.accept = fi_no_accept,
+	.reject = fi_no_reject,
+	.shutdown = fi_no_shutdown,
+};
+
+int smr_getopt(fid_t fid, int level, int optname,
+		void *optval, size_t *optlen)
+{
+	return -FI_ENOPROTOOPT;
+}
+
+int smr_setopt(fid_t fid, int level, int optname,
+		const void *optval, size_t optlen)
+{
+	return -FI_ENOPROTOOPT;
+}
+
+static struct fi_ops_ep smr_ep_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = fi_no_cancel,
+	.getopt = smr_getopt,
+	.setopt = smr_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void smr_tx_comp(struct smr_ep *ep, void *context)
+{
+	struct fi_cq_tagged_entry *comp;
+
+	comp = ofi_cirque_tail(ep->util_ep.tx_cq->cirq);
+	comp->op_context = context;
+	comp->flags = FI_SEND;
+	comp->len = 0;
+	comp->buf = NULL;
+	comp->data = 0;
+	ofi_cirque_commit(ep->util_ep.tx_cq->cirq);
+}
+
+static void smr_tx_comp_signal(struct smr_ep *ep, void *context)
+{
+	smr_tx_comp(ep, context);
+	ep->util_ep.tx_cq->wait->signal(ep->util_ep.tx_cq->wait);
+}
+
+static void smr_rx_comp(struct smr_ep *ep, void *context, uint64_t flags,
+			 size_t len, void *buf, void *addr)
+{
+	struct fi_cq_tagged_entry *comp;
+
+	comp = ofi_cirque_tail(ep->util_ep.rx_cq->cirq);
+	comp->op_context = context;
+	comp->flags = FI_RECV | flags;
+	comp->len = len;
+	comp->buf = buf;
+	comp->data = 0;
+	ofi_cirque_commit(ep->util_ep.rx_cq->cirq);
+}
+
+static void smr_rx_src_comp(struct smr_ep *ep, void *context, uint64_t flags,
+			     size_t len, void *buf, void *addr)
+{
+	ep->util_ep.rx_cq->src[ofi_cirque_windex(ep->util_ep.rx_cq->cirq)] =
+		(uint32_t) (uintptr_t) addr;
+	smr_rx_comp(ep, context, flags, len, buf, addr);
+}
+
+static void smr_rx_comp_signal(struct smr_ep *ep, void *context,
+			uint64_t flags, size_t len, void *buf, void *addr)
+{
+	smr_rx_comp(ep, context, flags, len, buf, addr);
+	ep->util_ep.rx_cq->wait->signal(ep->util_ep.rx_cq->wait);
+}
+
+static void smr_rx_src_comp_signal(struct smr_ep *ep, void *context,
+			uint64_t flags, size_t len, void *buf, void *addr)
+{
+	smr_rx_src_comp(ep, context, flags, len, buf, addr);
+	ep->util_ep.rx_cq->wait->signal(ep->util_ep.rx_cq->wait);
+
+}
+
+void smr_progress_resp(struct smr_ep *ep)
+{
+	struct smr_resp *resp;
+
+	fastlock_acquire(&ep->region->lock);
+	while (!ofi_cirque_isempty(smr_resp_queue(ep->region))) {
+		resp = ofi_cirque_head(smr_resp_queue(ep->region));
+		if (resp->status)
+			break;
+		ep->tx_comp(ep, NULL);
+		ofi_cirque_discard(smr_resp_queue(ep->region));
+	}
+	fastlock_release(&ep->region->lock);
+}
+
+void smr_progress_cmd(struct smr_ep *ep)
+{
+	struct smr_ep_entry *entry;
+	struct smr_cmd *cmd;
+	struct smr_resp *resp;
+	struct smr_inject_buf *tx_buf;
+	size_t inj_offset;
+	struct smr_region *peer_smr;
+	fi_addr_t addr;
+	int peer_id, ret = 0;
+
+	fastlock_acquire(&ep->region->lock);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
```

Comment:
Actually, see if we need the rc_cq lock, or if we can just use the region lock, since both locks are being held here, blocking rx cq processing anyway.